tools/simdgen/simdgen.ml

```diff
@@ -0,0 +1,373 @@
+      | "r16/m16" | "r/m16" -> [R16; M16]
+      | "r32/m32" | "r/m32" -> [R32; M32]
+      | "r64/m64" | "r/m64" -> [R64; M64]
+      | "r32/m8" -> [R32; M8]
+      | "r32/m16" -> [R32; M16]
+      | "r8" -> [R8]
+      | "r16" -> [R16]
+      | "r32" | "r32a" | "r32b" -> [R32]
+      | "r64" | "r64a" | "r64b" | "reg" -> [R64]
+      | "reg/m8" -> [R64; M8]
+      | "reg/m16" -> [R64; M16]
+      | "reg/m32" -> [R64; M32]
+      | "reg/m64" -> [R64; M64]
+      | "m8" -> [M8]
+      | "m16" -> [M16]
+      | "m32" -> [M32]
+      | "m64" -> [M64]
+      | "m128" -> [M128]
+      | "mm" | "mm0" | "mm1" | "mm2" | "mm3" -> [MM]
+      | "mm0/m8" | "mm1/m8" | "mm2/m8" | "mm3/m8" -> [MM; M8]
+      | "mm0/m16" | "mm1/m16" | "mm2/m16" | "mm3/m16" -> [MM; M16]
+      | "mm0/m32" | "mm1/m32" | "mm2/m32" | "mm3/m32" -> [MM; M32]
+      | "mm0/m64" | "mm1/m64" | "mm2/m64" | "mm3/m64" -> [MM; M64]
+      | "xmm" | "xmm0" | "xmm1" | "xmm2" | "xmm3" -> [XMM]
+      | "xmm0/m8" | "xmm1/m8" | "xmm2/m8" | "xmm3/m8" -> [XMM; M8]
+      | "xmm0/m16" | "xmm1/m16" | "xmm2/m16" | "xmm3/m16" -> [XMM; M16]
+      | "xmm0/m32" | "xmm1/m32" | "xmm2/m32" | "xmm3/m32" -> [XMM; M32]
+      | "xmm0/m64" | "xmm1/m64" | "xmm2/m64" | "xmm3/m64" -> [XMM; M64]
+      | "xmm0/m128" | "xmm1/m128" | "xmm2/m128" | "xmm3/m128" -> [XMM; M128]
+      | "mem" -> raise Unsupported
+      | _ -> failwith name
+    in
+    let enc, rw = first_word enc in
+    let enc =
+      match String.trim enc with
+      | "ModRM:reg" -> RM_r
+      | "ModRM:r/m" -> RM_rm
+      | "VEX.vvvv" -> Vex_v
+      | "imm8" | "imm16" | "imm32" | "imm64" | "Imm8" | "Imm16" | "Imm32"
+      | "Imm64" | "NA" | "<XMM0>" | "implicit" ->
+        Implicit
+      | _ -> failwith name
+    in
+    let is_imm = List.length locs = 0 in
+    let is_res =
+      match String.trim rw with
+      | "(w)" ->
+        set_res locs enc;
+        true
+      | _ -> false
+    in
+    let acc = if not (is_res || is_imm) then { locs; enc } :: acc else acc in
+    parse_args name acc encs args imm res
+  | _ -> failwith "mismatch"
+
+let parse_args name enc args =
+  let imm = ref None in
+  let res = ref None in
+  let args = parse_args name [] enc args imm res in
+  args, !imm, !res
+
+let implicit_result = function
+  | "pcmpistri" | "vpcmpistri" | "pcmpestri" | "vpcmpestri" ->
+    Fresh ([RCX], Implicit)
+  | "pcmpistrm" | "vpcmpistrm" | "pcmpestrm" | "vpcmpestrm" ->
+    Fresh ([XMM0], Implicit)
+  | _ -> First
+
+let implicit_args = function
+  | "pcmpestri" | "vpcmpestri" | "pcmpestrm" | "vpcmpestrm" ->
+    [{ locs = [RAX]; enc = Implicit }; { locs = [RDX]; enc = Implicit }]
+  | "mulx" -> [{ locs = [RDX]; enc = Implicit }]
+  | _ -> []
+
+let parse_ext = function
+  | "SSE" -> Some SSE
+  | "SSE2" -> Some SSE2
+  | "SSE3" -> Some SSE3
+  | "SSSE3" -> Some SSSE3
+  | "SSE4_1" -> Some SSE4_1
+  | "SSE4_2" -> Some SSE4_2
+  | "PCLMULQDQ" -> Some CLMUL
+  | "BMI2" -> Some BMI2
+  | _ -> None
+
+let print_sse42_extra () =
```

Comment:
why are these not in the csv?

Reply:
these aren't actually instructions; they compile to the string instruction + set/zext.
We shouldn't really represent them as instructions, but I don't think we have a way to encode the sequence in Cmm operations, since it relies on flags.

Reply:
added them to the csv manually

Reply:
fixed